interface/ceed-operator.c

```diff
@@ -298,6 +298,200 @@ int CeedOperatorAssembleLinearQFunction(CeedOperator op, CeedVector *assembled,
   return 0;
 }
 
+/**
+  @brief Assemble the diagonal of a linear Operator
+
+  This returns a CeedVector containing the diagonal of a linear CeedOperator.
+
+  @param op             CeedOperator to assemble CeedQFunction
+  @param[out] assembled CeedVector to store assembled CeedOperator diagonal
+  @param request        Address of CeedRequest for non-blocking completion, else
+                          CEED_REQUEST_IMMEDIATE
+
+  @return An error code: 0 - success, otherwise - failure
+
+  @ref Basic
+**/
+int CeedOperatorAssembleLinearDiagonal(CeedOperator op,
+    CeedVector *assembled, CeedRequest *request) {
+  int ierr;
+  Ceed ceed = op->ceed;
+  CeedQFunction qf = op->qf;
+
+  if (op->composite) {
+    // LCOV_EXCL_START
+    return CeedError(ceed, 1, "Cannot assemble QFunction for composite operator");
+  // LCOV_EXCL_STOP
+  } else {
+    if (op->nfields == 0)
+      // LCOV_EXCL_START
+      return CeedError(ceed, 1, "No operator fields set");
+    // LCOV_EXCL_STOP
+    if (op->nfields < qf->numinputfields + qf->numoutputfields)
+      // LCOV_EXCL_START
+      return CeedError( ceed, 1, "Not all operator fields set");
+    // LCOV_EXCL_STOP
+    if (op->numelements == 0)
+      // LCOV_EXCL_START
+      return CeedError(ceed, 1, "At least one restriction required");
+    // LCOV_EXCL_STOP
+    if (op->numqpoints == 0)
+      // LCOV_EXCL_START
+      return CeedError(ceed, 1, "At least one non-collocated basis required");
+    // LCOV_EXCL_STOP
+  }
+
+  // Use backend version, if available
+  if (op->AssembleLinearDiagonal) {
+    ierr = op->AssembleLinearDiagonal(op, assembled, request); CeedChk(ierr);
+    return 0;
+  }
```

Comment:
I currently don't have a good delegation mechanism for other backends to evaluate CeedOperatorAssembleLinearQF via ref. Once I get that figured out, it would make sense to put this ref and use a similar delegation so all backends get OperatorAssembleLinearDiag for free.